src/libsyntax/parse/parser.rs

```diff
@@ -1282,12 +1290,19 @@ impl<'a> Parser<'a> {
                     Err(e) => {
                         loop {
                             p.bump();
-                            if p.token == token::Semi {
+
+                            // Don't bump past the EOF.
+                            if p.token == token::Eof {
```

Comment:
To me it feels like at this point converting this chain to a `match` would be better here.